design/decisions/0005-entity-registration-api.md

```diff
@@ -0,0 +1,398 @@
+  "entities": [
+      {
+          "@topic-id": "device/child01//",
+          "@type": "child-device",
+          "@id": "child01",
+          "@parent": "device/main//",
+          "name": "child01",
+          "type": "Raspberry Pi",
+          "extra-fragment": {
+              "extra-key": "extra-value"
+          },
+          "new-fragment": {
+              "new-key": "new-value"
+          }
+      },
+      {
+          "@topic-id": "device/child02//",
+          "@type": "child-device",
+          "@parent": "device/main//",
+          "@id": "child02",
+          "name": "child02",
+          "type": "Raspberry Pi",
+      },
+      {
+          "@topic-id": "device/main/service/service01",
+          "@type": "service",
+          "@parent": "device/main//",
+          "@id": "service01",
+          "name": "service01"
+      },
+      ...
+  ],
+  "page": 1,
+  "pageSize": 10,
+  "totalPages": 7
+  ```
+
+:::warning
+Supporting filtering queries and paginated results would require the `tedge-agent` to maintain
+additional data structures like indexes and other things which are yet to be explored/defined.
+:::
+
+#### Nested queries for immediate and nested child entities of an entity
+
+Recursive query of immediate and nested children of `device/main//`:
+
+```
+GET /v1/entities?%40parent=device/main//&recursive=true
+```
+
+Immediate and nested children of `device/main//`, with a `depth` limit:
+
+```
+GET /v1/entities?%40parent=device/main//&recursive=true&depth=2
+```
+
+**Responses**
+
+* 200: OK
+  ```json
+  "entities": [
+      {
+          "@topic-id": "device/child01//",
+          "@type": "child-device",
+          "@id": "child01",
+          "@parent": "device/main//",
+          "name": "child01",
+          "type": "Raspberry Pi",
+          "extra-fragment": {
+              "extra-key": "extra-value"
+          },
+          "new-fragment": {
+              "new-key": "new-value"
+          }
+      },
+      {
+          "@topic-id": "device/child02//",
+          "@type": "child-device",
+          "@parent": "device/main//",
+          "@id": "child02",
+          "name": "child02",
+          "type": "Raspberry Pi",
+      },
+      {
+          "@topic-id": "device/main/service/service01",
+          "@type": "service",
+          "@parent": "device/main//",
+          "@id": "service01",
+          "name": "service01"
+      },
+      {
+          "@topic-id": "device/child01/service/service01",
+          "@type": "service",
+          "@parent": "device/child01//",
+          "@id": "service01",
+          "name": "service01"
+      },
+      {
+          "@topic-id": "device/child02/service/service01",
+          "@type": "service",
+          "@parent": "device/child02//",
+          "@id": "service01",
+          "name": "service01"
+      }
+  ]
+  ```
+
+#### Delete all child entities of an entity
+
+Delete all children of `device/child01//`:
+
+```
+DELETE /v1/entities?%40parent=device/child01//
```

Comment:
This is slightly unconventional, to `DELETE` a resource using query parameters. The conventional way would have been to use sub paths of the target resource:

```
DELETE /v1/entities/{topic-id}/children
```

But, I chose to avoid sub-paths due to the empty `/` characters in the `topic-id` as follows:

```
DELETE /v1/entities/device/child01///children
```

...which is not an acceptable convention either.

Another alternative would have been to provide a dedicated path with the target entity `topic-id` provided as a query parameter as follows:

```
DELETE /v1/entities/children?%40topic-id=device/child01//
```

But, I don't find it to be much of an improvement from the existing version either.

Reply:
Do we really need a way to `DELETE` in one operation all the child-devices of given device?

Isn't a  child-devices query and a series one-by-one `DELETE`s enough?

Reply:
> Do we really need a way to `DELETE` in one operation all the child-devices of given device?
> 
> Isn't a child-devices query and a series of one-by-one `DELETE`s enough?

I would agree with @didier-wenzek. We don't need an explicit "delete all children" API command as we can do a query and then do individual api calls (we could also wrap the calls in a special tedge cli command to still do the same function but not require a specific API call).

Reply:
Addressed by eb0eba88aa10427f4eb7ad70a0c04677202395b5